### PR TITLE
Fix segfault: Do not combine last turn channel maneuver

### PR DESCRIPTION
**Pull request by danpaz**
# Issue

Fixes an invalid access bug triggered by combining maneuvers at a turn channel that is also a roundabout exit. The fix is to simply skip the turn channel collapsing logic if the next maneuver is the last maneuver in the route.

Although the turn channel collapsing logic is not new and the bug has been in valhalla for a while, the segfault actually occurs further along in new code that was added to collapse roundabout exit maneuvers. The combination of collapsing the final maneuver, then collapsing again for roundabout_exits=false, triggers the bug.

Address sanitizer output: [asan.txt](https://github.com/valhalla/valhalla/files/4905270/asan.txt)

We haven't been able to craft a reproducible test case in gurka yet as it's been tricky getting the tags just right for a link to be marked as a turn_channel. I'll follow up with a test case in a future PR.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/odin/maneuversbuilder.cc [hunk 1]
```diff
@@ -2336,7 +2336,7 @@ bool ManeuversBuilder::IsTurnChannelManeuverCombinable(std::list<Maneuver>::iter
                                                        bool start_man) const {
 
   // Current maneuver must be a turn channel and not equal to the next maneuver
-  if (curr_man->turn_channel() && (curr_man != next_man)) {
+  if (curr_man->turn_channel() && (curr_man != next_man) && !next_man->IsDestinationType()) {
```
Comment (mandeepsandhu) [inline src/odin/maneuversbuilder.cc:2339]: Should we leave a TODOD note that we need to extend this logic (of not combining with the destination maneuver) to other places as well?

Are there tests which exercise this code-path when roundabout_exits is true & false?

Comment (danpaz) [reply src/odin/maneuversbuilder.cc:2339]: I'm actually not sure where it makes sense to extend this logic, as it seems to have been deliberately left out of the other collapse cases. I'll follow up with @dgearhart about that as soon as I can.

Re: tests, I mentioned in the OP that I'll follow up with a gurka test. Meanwhile @gknisely helped run RAD tests with this change and saw no diffs, so I'm fairly confident the fix is localized to the situation that triggers the bug.

Comment (mandeepsandhu) [reply src/odin/maneuversbuilder.cc:2339]: Okay. RAD showing 0 diffs is a good sign! :) 👍

**Comment by danpaz** [comment]
@gknisely thank you for adding the gurka test! I confirmed it fails on master due to missing the destination maneuver:
```
Expected equality of these values:
  actual_maneuvers
    Which is: { 1, 20, 26, 27 }
  expected_maneuvers
    Which is: { 1, 20, 26, 27, 5 }
Actual maneuvers didn't match expected maneuvers
```

File: test/gurka/test_instructions_roundabout.cc [hunk 2]
```diff
@@ -203,3 +203,72 @@ TEST_F(InstructionsRoundabout, RoundaboutExitSuppressed) {
...
+  const double gridsize = 10;
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, gridsize, {5.1079374, 52.0887174});
+
+  auto map = gurka::buildtiles(layout, ways, nodes, {}, "test/data/gurka_roundabouts_regression",
+                               {{"mjolnir.admin",
+                                 {VALHALLA_SOURCE_DIR "test/data/netherlands_admin.sqlite"}}});
+
+  const std::string& request =
+      (boost::format(
+           R"({"locations":[{"lat":52.08862730752708,"lon":5.107829332318943},{"lat":52.08843613199833,"lon":5.109036862850189}],"costing":"auto"})"))
```
Comment (mandeepsandhu) [inline test/gurka/test_instructions_roundabout.cc:251]: @gknisely how did you determine the lat/lon of the request? Is it possible to specify the source & destination using their alphabet strings (and then use `map.nodes.at(from).lat())` & `map.nodes.at(from).lng())` ) instead? I think this src/dest in this test are `K -> P` ?

Comment (danpaz) [reply test/gurka/test_instructions_roundabout.cc]: @mandeepsandhu I updated with input points `1` and `2` to clarify.

**Pull request merged**